Description: Fixes a regression causing the encountered/expected token list to be empty
Author: Dag H. Wanvik <dag.wanvik@oracle.com>
Forwarded: yes
Bug: https://java.net/jira/browse/JAVACC-244
--- a/src/org/javacc/parser/ParseGen.java
+++ b/src/org/javacc/parser/ParseGen.java
@@ -570,21 +570,24 @@
           ostr.println("      for (int i = 0; i < jj_endpos; i++) {");
           ostr.println("        jj_expentry[i] = jj_lasttokens[i];");
           ostr.println("      }");
+          ostr.println("      boolean exists = false;");
         if (!Options.getGenerateGenerics())
-          ostr.println("      jj_entries_loop: for (java.util.Iterator it = jj_expentries.iterator(); it.hasNext();) {");
+          ostr.println("      for (java.util.Iterator it = jj_expentries.iterator(); it.hasNext();) {");
         else
-          ostr.println("      jj_entries_loop: for (java.util.Iterator<?> it = jj_expentries.iterator(); it.hasNext();) {");
+          ostr.println("      for (java.util.Iterator<?> it = jj_expentries.iterator(); it.hasNext();) {");
+          ostr.println("        exists = true;");
           ostr.println("        int[] oldentry = (int[])(it.next());");
           ostr.println("        if (oldentry.length == jj_expentry.length) {");
           ostr.println("          for (int i = 0; i < jj_expentry.length; i++) {");
           ostr.println("            if (oldentry[i] != jj_expentry[i]) {");
-          ostr.println("              continue jj_entries_loop;");
+          ostr.println("              exists = false;");
+          ostr.println("              break;");
           ostr.println("            }");
           ostr.println("          }");
-          ostr.println("          jj_expentries.add(jj_expentry);");
-          ostr.println("          break jj_entries_loop;");
+          ostr.println("          if (exists) break;");
           ostr.println("        }");
           ostr.println("      }");
+          ostr.println("      if (!exists) jj_expentries.add(jj_expentry);");
           ostr.println("      if (pos != 0) jj_lasttokens[(jj_endpos = pos) - 1] = kind;");
           ostr.println("    }");
           ostr.println("  }");
